File: python_version/linked-lists/reorder_list.py
class Solution:
    """
    @param head: The head of linked list.
    @return: nothing
    """

    def findMiddle(self, h):
        slow, fast = h, h.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow

    def reverse(self, h):
        prev = None
        while h:
            tmp = h.next
            h.next = prev
            prev = h
            h = tmp

        return prev

    def reorderList(self, head):
        # write your code here
        if head is None or head.next is None:
            return head

        mid = self.findMiddle(head)
        last = self.reverse(mid.next)
        mid.next = None

        while last:
            tmp = head.next
            head.next = last
            h = last.next
            last.next = tmp
            head = tmp
            last = h

File: src/history/linked-lists/reorder-list.java
class Solution {
    public void reorderList(ListNode head) {
        ListNode slow = head, fast = head;
        while(fast != null && fast.next != null){
          slow = slow.next;
          fast = fast.next.next;
        }
        ListNode reverseLastHalf = reverse(slow);
        ListNode firstHalf = head;
        while(firstHalf != null && reverseLastHalf != null){
          ListNode headTemp = firstHalf.next;
          firstHalf.next = reverseLastHalf;
          firstHalf = headTemp;

          ListNode reverseTemp = reverseLastHalf.next;
          reverseLastHalf.next = firstHalf;
          reverseLastHalf = reverseTemp;
        }
        if(firstHalf != null){
          firstHalf.next = null;
        }
    }

    public static ListNode reverse(ListNode head){
        ListNode prev = null, temp = null;
        while(head != null){
          temp = head;
          head = head.next;
          temp.next = prev;
          prev = temp;
        }
        return prev;
    }
}